multilinear_extensions/src/smart_slice.rs

```diff
@@ -0,0 +1,167 @@
+use serde::{Deserialize, Deserializer, Serialize};
+use std::{
+    hash::{Hash, Hasher},
+    ops::{Deref, DerefMut},
+};
+
+/// a flexible enum similar to `std::borrow::Cow`, but supports mutable slices as a variant
+///
+/// `std::Cow` allows either borrowed (`&T`) or owned (`T::Owned`) data, but it does **not**
+/// support mutable borrowed data (i.e., `&mut [T]`). This `SmartSlice` enum extends Cow-like
+/// behavior by adding a third variant:
+/// - `Borrowed`: a shared reference to `[T]`, like `Cow::Borrowed`
+/// - `BorrowedMut`: a mutable reference to `[T]`, not supported in `Cow`
+/// - `Owned`: an owned `Vec<T>`, like `Cow::Owned`
+///
+/// this is useful when the caller might own, share, or temporarily mutate a slice
+/// and we want to abstract over all three cases while enabling mutation (via `.to_mut()`).
+#[derive(Debug, Serialize)]
+pub enum SmartSlice<'a, T> {
+    Borrowed(&'a [T]),        // shared reference
+    BorrowedMut(&'a mut [T]), // mutable reference
+    Owned(Vec<T>),            // owned vector
+}
+
+impl<'a, T> Default for SmartSlice<'a, T> {
+    fn default() -> Self {
+        SmartSlice::Owned(Vec::new()) // default to empty owned vector
+    }
+}
+
+impl<'a, T> SmartSlice<'a, T> {
+    /// ensures the data is owned and returns a mutable slice.
+    ///
+    /// Panic if the data is reference borrowed
+    pub fn to_mut(&mut self) -> &mut [T]
+    where
+        T: Clone,
+    {
+        match self {
+            SmartSlice::Borrowed(_) => unimplemented!("calling to_mut on immutable slice"),
```

Comment:
nitpick: I think the right macro to use is `panic!()` not `unimplemented!`. It's considered a bug if someone called `to_mut` on a immutable borrowed slice.